sdk/go/pulumi/context.go

```diff
@@ -177,15 +183,26 @@ func (ctx *Context) Context() context.Context {
 	return ctx.ctx
 }
 
+// Returns a copy of base context in which the value associated with key is val.
+func (ctx *Context) WithValue(key, val any) *Context {
+	newCtx := &Context{
+		ctx:   ctx.ctx,
+		state: ctx.state,
+		Log:   ctx.Log,
+	}
+	newCtx.ctx = context.WithValue(newCtx.ctx, key, val)
+	return newCtx
+}
```

Comment:
I wonder if we should also expose a `Value(...)` function here.  Right now it is exposed through `c.Context().Value(...)`, but it feels a little bit unintuitive to me that the user needs knowledge that this is internally stored on the Go context if we're exposing the `WithValue` function here.  Curious what others think.

Reply:
I think that would be good. If @emilhotkowski is happy to add it to this PR great, it's small enough that I don't think it really increases the review burden. Else we can add it later.

Reply:
I agree this makes for cleaner API. I'll add it.

Reply:
Done